libraries/AP_OSD/AP_OSD_Screen.cpp

```diff
@@ -1917,6 +2023,192 @@ void AP_OSD_Screen::draw_esc_amps(uint8_t x, uint8_t y)
+{
+    const int8_t snr = AP::crsf()->get_link_status().snr;
+    const bool blink = snr < osd->warn_snr;
+    if(this->is_btfl_fonts()) {
+        this->draw_crsf_snr_btfl(x, y, snr, blink);
+    }
+    else {
+        if (snr == INT8_MIN) {
+            backend->write(x, y, blink, "%c---%c", SYMBOL(SYM_SNR), SYMBOL(SYM_DB));
+        }
+        else {
+            backend->write(x, y, blink, "%c%3d%c", SYMBOL(SYM_SNR), snr, SYMBOL(SYM_DB));
+        }
+    }    
+}
+
+void AP_OSD_Screen::draw_crsf_snr_btfl(uint8_t x, uint8_t y, int8_t value, bool blink)
+{    
+    if (value == INT8_MIN) {
+        backend->write(x, y, blink, "SNR---DB");
+    }
+    else {
+        backend->write(x, y, blink, "SNR%3dDB", value);
+    }
+}
+
+void AP_OSD_Screen::draw_crsf_active_antenna(uint8_t x, uint8_t y)
+{
+    const int8_t active_antenna = AP::crsf()->get_link_status().active_antenna;
+    if(this->is_btfl_fonts()) {
+        this->draw_crsf_active_antenna_btfl(x, y, active_antenna);
+    }
+    else {
+        if (active_antenna < 0) {
+            backend->write(x, y, false, "%c-", SYMBOL(SYM_ANT));
+        } 
+        else {
+            backend->write(x, y, false, "%c%d", SYMBOL(SYM_ANT), active_antenna + 1);
+        }
+    }    
+}
+
+void AP_OSD_Screen::draw_crsf_active_antenna_btfl(uint8_t x, uint8_t y, int8_t value)
+{    
+    if (value < 0) {
+        backend->write(x, y, false, "ANT-");
+    } else {
+        backend->write(x, y, false, "ANT%d", value + 1);
+    }
+}
+
+void AP_OSD_Screen::draw_crsf_lq(uint8_t x, uint8_t y)
+{    
+    const int16_t lqv = AP::crsf()->get_link_status().link_quality;
+    const bool blink = lqv < osd->warn_lq;
+    if(this->is_btfl_fonts()) {
+        this->draw_crsf_lq_btfl(x, y, lqv, blink);
+    }
+    else {
+        if(check_option(AP_OSD::OPTION_RF_MODE_ALONG_WITH_LQ)) {
+            const int16_t rf_mode = AP::crsf()->get_link_status().rf_mode;        
+            if (lqv < 0) {            
+                backend->write(x, y, blink, "%c--:--", SYMBOL(SYM_LQ));
+            }        
+            else {            
+                backend->write(x, y, blink, "%c%2d:%2d", SYMBOL(SYM_LQ), rf_mode, lqv);
+            }
+        }        
+        else {           
+            if (lqv < 0) {            
+                backend->write(x, y, blink, "%c--", SYMBOL(SYM_LQ));
+            }
+            else {            
+                backend->write(x, y, blink, "%c%2d", SYMBOL(SYM_LQ), lqv);
+            }        
+        }
+    }
+}
+
+void AP_OSD_Screen::draw_crsf_lq_btfl(uint8_t x, uint8_t y, int16_t value, bool blink)
+{
+    if(check_option(AP_OSD::OPTION_RF_MODE_ALONG_WITH_LQ)) {
+        const int16_t rf_mode = AP::crsf()->get_link_status().rf_mode;        
+        if (value < 0) {            
+            backend->write(x, y, blink, "LQ--:--");
+        }        
+        else {            
+            backend->write(x, y, blink, "LQ%2d:%2d", rf_mode, value);
```

Comment:
A chunk of the flash cost is all these very similar functions. I think you should be able to boil this down to a lookup table and a much smaller set of functions

Reply:
Indeed! I have rearranged these functions (in commit [d419b69](https://github.com/ArduPilot/ardupilot/pull/25844/commits/d419b699c89347baeca8474349063f18be8caa4f)) to the best I could. A lookup table isn't practical due to the variadic arguments in the calls to `backend->write()` and the requirement of the format string to be a `const char*`. I couldn't think of any solution that made sense other than the traditional "if else" structuring. But they are now all contained inside the relevant functions, without duplicates. Saved about 200 bytes of flash.